---
title: "CRISPR Report"
output: html_document
date: '`r date()`'
params: 
  sampleSheetFile: ''
  ampliconFastaFile: ''
  ampliconName: ''
  cutSitesFile: ''
  bedgraphFolder: ''
  workdir: ''
  prefix: ''
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
knitr::opts_knit$set(root.dir = params$workdir)
library(knitr)
library(ggplot2)
library(ggrepel)
library(data.table)
library(rtracklayer)
library(DT)
library(plotly)
library(crosstalk)
```

# Input Settings
```{r printInputSettings}
ampliconFastaFile <- params$ampliconFastaFile
ampliconName <- params$ampliconName
sampleSheetFile <- params$sampleSheetFile
bedgraphFolder <- params$bedgraphFolder
cutSitesFile <- params$cutSitesFile
prefix <- params$prefix
workdir <- params$workdir

inputParameterDesc <- c('Amplicon Fasta File',
                        'Amplicon Name',
                        'sgRNA Cut Sites File', 
                     'Experiment Data File',
                     'bedgraph folder path', 
                     'Prefix for output files',
                     'Working directory'
                     )
inputParameterValues <- c(ampliconFastaFile,
                          ampliconName,
                          cutSitesFile, 
                          sampleSheetFile,
                          bedgraphFolder, 
                          prefix,
                          workdir)
inputSettings <- data.frame(parameters = inputParameterDesc,
                            values = inputParameterValues,
                            stringsAsFactors = FALSE)
DT::datatable(data = inputSettings,
              extensions = 'FixedColumns',
              options = list(fixedColumns = TRUE,
                         scrollX = TRUE,
                         pageLength = 9,
                         dom = 't'))
```

```{r printSampleInformation}
sampleSheet <- read.csv(sampleSheetFile, stringsAsFactors = F)
sampleSheet <- sampleSheet[sampleSheet$amplicon == ampliconName,]
DT::datatable(data = sampleSheet,
              extensions = 'FixedColumns',
              options = list(fixedColumns = TRUE,
                         scrollX = TRUE,
                         dom = 't'))

cutSites <- read.table(cutSitesFile, stringsAsFactors = F)
colnames(cutSites) <- c('sgRNA', 'cutSite')
DT::datatable(data = cutSites)

```

# Coverage Profiles 

## Coverage as Heatmap
```{r coverage_heatmap}
# import stats tables 
coverageStats <- as.data.table(do.call(rbind, lapply(sampleSheet$sample_name, function(sampleName) {
  f <- file.path(bedgraphFolder, paste0(sampleName, '.coverageStats.tsv'))
  if(file.exists(f)) {
    dt <- data.table::fread(f)
    return(dt)
  }
})))
#create a matrix of bp versus samples where values are coverage
dt <- dcast.data.table(coverageStats, sample ~ bp, value.var = 'cov', fill = 0)
M <- as.matrix(dt[,-1])
rownames(M) <- dt$sample
pheatmap::pheatmap(M, 
                   cluster_rows = nrow(M) > 1, 
                   cluster_cols = F, 
                   show_colnames = F, 
                   main = paste(ampliconName, 'Coverage Profiles'))
```

## Coverage as Line Plot

```{r coverage-lineplot}
p <- ggplot(coverageStats, aes(x = bp, y = cov, group = sample), height = 500) + 
  geom_line(aes(color = sample), show.legend = F) +   
  theme_bw() +  
  labs(x = 'base position', y = 'Number of Reads', title = paste(ampliconName, 'Coverage Profiles'))

ggplotly(p)
```

# Indel Score Profiles 

```{r plotIndelProfiles}

plotScores <- function(bedgraphFolder, sampleName, sampleSheet, cutSites) {
  f <- file.path(bedgraphFolder, paste0(sampleName, '.coverageStats.tsv'))
  dt <- data.table::fread(f)

  sgRNAs <- unlist(strsplit(x = sampleSheet[sampleSheet$sample_name == sampleName,]$sgRNA_ids, 
                     split = ':'))
  
  sgRNAs <- merge(data.frame('sgRNA' = sgRNAs, 'sample_name' = sampleName), cutSites, by = 'sgRNA')

  mdt <- melt(dt[,-c('ins', 'del', 'indel', 'cov')], id.vars = c('bp', 'sample', 'seqname'))
  
  p <- ggplot(mdt, aes(x = bp, y = value, group = variable)) +
    geom_line(aes(color = variable)) + labs(title = sampleName)
  if(nrow(sgRNAs) > 0) {
    p <- p + geom_vline(data = sgRNAs, aes(xintercept = cutSite, 
                                           color = sgRNA), linetype = 'dotted')
  }
  p <- p + 
    theme_bw(base_size = 14) + 
    scale_y_continuous(labels = scales::percent)
  return(p)
}

plots <- lapply(sampleSheet$sample_name, function(sampleName) {
  plotScores(bedgraphFolder = bedgraphFolder, 
             sampleName = sampleName, 
             sampleSheet = sampleSheet, 
             cutSites = cutSites)
})
names(plots) <- sampleSheet$sample_name
```

```{r plotIndelProfilesHeader, results = 'asis'}
if(length(plots) > 10) {
  cat("## Indel Score Profiles {.tabset .tabset-dropdown}\n\n")
} else {
  cat("## Indel Score Profiles {.tabset}\n\n")
}
```

```{r plotIndelProfilesPlots}
out = NULL
for (n in names(plots)) {
  p <- ggplotly(plots[[n]])
  out = c(out, knitr::knit_expand(text='### {{n}} \n\n {{p}} \n\n'))
}
```

`r paste(knit(text = out), collapse = '\n')`

# sgRNA efficiencies at cut sites
```{r cutSiteStats}
# this is a table for all samples versus all cut sites in the amplicon
# not all sample x sgRNA combinations are true, 
# but still they are also computed to serve as a negative control
cutSiteStats <- as.data.table(do.call(rbind, lapply(sampleSheet$sample_name, function(sampleName) {
  f <- file.path(bedgraphFolder, paste0(sampleName, '.indel_stats_at_cutsites.tsv'))
  if(file.exists(f)) {
    dt <- fread(f)
    return(dt)
  }
})))

# match samples to the actual sgRNA guides used in that sample
sampleGuides <- lapply(sampleSheet$sample_name, function(s) {
  sgRNAs <- unlist(strsplit(x = sampleSheet[sampleSheet$sample_name == s,]$sgRNA_ids, 
                   split = ':'))
})
names(sampleGuides) <- as.character(sampleSheet$sample_name)

cutSiteStats$sampleMatchesGuide <- as.factor(apply(cutSiteStats, 1, function(x) {
  s <- as.character(x[['sample']])
  g <- as.character(x[['sgRNA']])
  return(g %in% sampleGuides[[s]])
}))

plots <- lapply(as.vector(unique(cutSiteStats$sgRNA)), function(sg) {
  dt <- cutSiteStats[sgRNA == sg & sampleMatchesGuide == TRUE]
  if(nrow(dt) > 0) {
    p <- ggplot(dt, aes(x = sample, y = indelEfficiency)) + 
      geom_bar(stat = 'identity', aes(fill = sample)) + 
      labs(title = sg) + coord_flip()
    return(p)
  }
})
names(plots) <- as.vector(unique(cutSiteStats$sgRNA))
```


```{r plotIndelStatsHeader, results = 'asis'}
if(length(plots) > 10) {
  cat("## Indel Stats At Cut Sites {.tabset .tabset-dropdown}\n\n")
} else {
  cat("## Indel Stats At Cut Sites {.tabset}\n\n")
}
```

```{r plotIndelStatsPlots}
out = NULL
for (n in names(plots)) {
  if(!is.null(plots[[n]])) {
  p <- ggplotly(plots[[n]])
  out = c(out, knitr::knit_expand(text='### {{n}} \n\n {{p}} \n\n'))
  } else {
    out = c(out, knitr::knit_expand(text='### {{n}} \n\n No indel data found\n\n'))
  }
}
```

`r paste(knit(text = out), collapse = '\n')`

# sgRNA efficiencies at cut sites - version 2

```{r}
dts <- crosstalk::SharedData$new(cutSiteStats)

bscols(widths = c(4,8),
  list(
    filter_checkbox("sampleMatchesGuide", "Own Guide", dts, ~factor(sampleMatchesGuide), inline = TRUE),
    filter_select( "sgRNA", "Select sgRNA", dts, ~factor(sgRNA), allLevels = FALSE, multiple = FALSE),
    filter_select( "sample", "Select samples", dts, ~factor(sample), allLevels = FALSE, multiple = TRUE)
    ),
  plotly::plot_ly(data = dts,
                  x = ~ sgRNA,
                  y = ~ indelEfficiency,
                  group = ~sgRNA,
                  color = ~sample,
                  type = 'bar',
                  text = ~sample,
                  height = 500)  %>%
    layout(showlegend = FALSE, xaxis = list(showticklabels = TRUE), barmode = 'group')
)
```


