---
title: "CRISPR Report"
output: html_document
date: '`r date()`'
params: 
  sampleSheetFile: ''
  ampliconFastaFile: ''
  ampliconName: ''
  cutSitesFile: ''
  bedgraphFolder: ''
  workdir: ''
  prefix: ''
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
knitr::opts_knit$set(root.dir = params$workdir)
library(knitr)
library(ggplot2)
library(ggrepel)
library(data.table)
library(rtracklayer)
library(DT)
library(plotly)
```

# Input Settings
```{r printInputSettings}
ampliconFastaFile <- params$ampliconFastaFile
ampliconName <- params$ampliconName
sampleSheetFile <- params$sampleSheetFile
bedgraphFolder <- params$bedgraphFolder
cutSitesFile <- params$cutSitesFile
prefix <- params$prefix
workdir <- params$workdir

inputParameterDesc <- c('Amplicon Fasta File',
                        'Amplicon Name',
                        'sgRNA Cut Sites File', 
                     'Experiment Data File',
                     'bedgraph folder path', 
                     'Prefix for output files',
                     'Working directory'
                     )
inputParameterValues <- c(ampliconFastaFile,
                          ampliconName,
                          cutSitesFile, 
                          sampleSheetFile,
                          bedgraphFolder, 
                          prefix,
                          workdir)
inputSettings <- data.frame(parameters = inputParameterDesc,
                            values = inputParameterValues,
                            stringsAsFactors = FALSE)
DT::datatable(data = inputSettings,
              extensions = 'FixedColumns',
              options = list(fixedColumns = TRUE,
                         scrollX = TRUE,
                         pageLength = 9,
                         dom = 't'))
```

```{r printSampleInformation}
sampleSheet <- read.csv(sampleSheetFile, stringsAsFactors = F)
sampleSheet <- sampleSheet[sampleSheet$amplicon == ampliconName,]
DT::datatable(data = sampleSheet,
              extensions = 'FixedColumns',
              options = list(fixedColumns = TRUE,
                         scrollX = TRUE,
                         dom = 't'))

cutSites <- read.table(cutSitesFile, stringsAsFactors = F)
colnames(cutSites) <- c('sgRNA', 'cutSite')
DT::datatable(data = cutSites)

```

# Coverage Profiles 

## Coverage as Heatmap
```{r coverage_heatmap}
# import stats tables 
coverageStats <- as.data.table(do.call(rbind, lapply(sampleSheet$sample_name, function(sampleName) {
  f <- file.path(bedgraphFolder, paste0(sampleName, '.coverageStats.tsv'))
  if(file.exists(f)) {
    dt <- data.table::fread(f)
    return(dt)
  }
})))
#create a matrix of bp versus samples where values are coverage
dt <- dcast.data.table(coverageStats, sample ~ bp, value.var = 'cov', fill = 0)
M <- as.matrix(dt[,-1])
rownames(M) <- dt$sample
pheatmap::pheatmap(M, 
                   cluster_rows = nrow(M) > 1, 
                   cluster_cols = F, 
                   show_colnames = F, 
                   main = paste(ampliconName, 'Coverage Profiles'))
```

## Coverage as Line Plot

```{r coverage-lineplot}
p <- ggplot(coverageStats, aes(x = bp, y = cov, group = sample)) + 
  geom_line(aes(color = sample), show.legend = F) +   
  theme_bw() +  
  labs(x = 'base position', y = 'Number of Reads', title = paste(ampliconName, 'Coverage Profiles'))
ggplotly(p)
```

# Indel Score Profiles 

```{r plotIndelProfiles}

plotScores <- function(bedgraphFolder, sampleName, sampleSheet, cutSites) {
  f <- file.path(bedgraphFolder, paste0(sampleName, '.coverageStats.tsv'))
  dt <- data.table::fread(f)

  sgRNAs <- unlist(strsplit(x = sampleSheet[sampleSheet$sample_name == sampleName,]$sgRNA_ids, 
                     split = ':'))
  
  sgRNAs <- merge(data.frame('sgRNA' = sgRNAs, 'sample_name' = sampleName), cutSites, by = 'sgRNA')

  mdt <- melt(dt[,-c('ins', 'del', 'indel', 'cov')], id.vars = c('bp', 'sample'))
  
  p <- ggplot(mdt, aes(x = bp, y = value, group = variable)) +
    geom_line(aes(color = variable)) + labs(title = paste(sampleName, 'Indel Profiles'))
  if(nrow(sgRNAs) > 0) {
    p <- p + geom_vline(data = sgRNAs, aes(xintercept = cutSite, 
                                           color = sgRNA), linetype = 'dotted')
  }
  p <- p + 
    theme_bw(base_size = 14) + 
    scale_y_continuous(labels = scales::percent)
  return(p)
}

plots <- lapply(sampleSheet$sample_name, function(sampleName) {
  plotScores(bedgraphFolder = bedgraphFolder, 
             sampleName = sampleName, 
             sampleSheet = sampleSheet, 
             cutSites = cutSites)
})
names(plots) <- sampleSheet$sample_name
```

```{r plotIndelProfilesHeader, results = 'asis'}
if(length(plots) > 10) {
  cat("## Indel Score Profiles {.tabset .tabset-dropdown}\n\n")
} else {
  cat("## Indel Score Profiles {.tabset}\n\n")
}
```

```{r plotIndelProfilesPlots, include=FALSE}
out = NULL
for (n in names(plots)) {
  p <- ggplotly(plots[[n]])
  out = c(out, knitr::knit_expand(text='### {{n}} \n\n {{p}} \n\n'))
}
```

`r paste(knit(text = out), collapse = '\n')`


